### ISSUE TITLE
Can't boot when using AHCI with VMWare Player

### ISSUE THREAD

**Issue by supercomputer7** [SerenityOS/serenity#5805]
As indicated by @Lubrsi, we can't yet boot with SATA (AHCI) drives with VMWare Player.

This can be either a problem in the handling code, but there's another possibility - we don't handle BIOS/OS Handoff Control,
so if VMWare BIOS requires us to do handle that, it's quite reasonable to fail when this is not handled correctly.

Let's get the basic details, so we need to answer a few questions:
1. What is the supported AHCI revision on VMWare Player?
2. How many ports it has for each controller?
3. Is it supporting AHCI power management capabilities?
4. Is it supporting Port-multiplier functionality?

**Comment by Lubrsi** [comment]
I think I've found out why. VMware Player is using IRQ 7 for the AHCI controller, which we reserve as a spurious interrupt handler.
This is probably causing us to skip the reset sequence for the ports.

VirtualBox:
```
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port Handler: IRQ 11
...snip...
[34;1m[init_stage2(1:1)][0m: AHCI Port 1: Device found, Capacity=494927872, Bytes per logical sector=512, Bytes per physical sector=512
[34;1m[#0 colonel(0:0)][0m: AHCI Port Handler: IRQ received
[34;1m[#0 colonel(0:0)][0m: AHCI Port Handler: Handling IRQ for port 0
[34;1m[#0 colonel(0:0)][0m: AHCI Port 1: Interrupt handled, PxIS 1
```

VMware:
```
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port Handler: IRQ 7
...snip...
[34;1m[init_stage2(1:1)][0m: AHCI Port 1: Device found, Capacity=494927872, Bytes per logical sector=512, Bytes per physical sector=512
[34;1m[#0 init_stage2(1:1)][0m: Spurious interrupt, vector 7
```

From the logs, before VB and VMW trigger an interrupt at the exact same time for AHCI. However, we throw away the interrupt as "spurious".

**Comment by Lubrsi** [comment]
Nevermind, I removed the spurious interrupt handler on IRQ 7 and it didn't fix the issue. At least interrupts come through though.

The PxIS for the first request don't seem to match:
VirtualBox:
```
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port 1: Do a read, lba 0, block count 1 @ P0x00e50000, ended
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port 1: Request success
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port Handler: IRQ received
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port Handler: Handling IRQ for port 0
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port 1: Interrupt handled, PxIS 1
```

VMW:
```
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port 1: Do a read, lba 0, block count 1 @ P0x00e02000, ended
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port 1: Request success
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port Handler: IRQ received
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port Handler: Handling IRQ for port 0
[34;1m[#0 init_stage2(1:1)][0m: AHCI Port 1: Interrupt handled, PxIS 1073741824
```

**Comment by Lubrsi** [comment]
These are the capabilities:
```
[34;1m[#0 init_stage2(1:1)][0m: PCI [0000:02:03:00]: AHCI Controller Capabilities = 0xc1341f1d
```

So 30 ports, Gen 3 (not sure about revision), doesn't support power management and doesn't support port multipliers.

**Comment by supercomputer7** [comment]
> I think I've found out why. VMware Player is using IRQ 7 for the AHCI controller, which we reserve as a spurious interrupt handler.
> This is probably causing us to skip something.
> 
> VirtualBox:
> 
> ```
> �[34;1m[#0 init_stage2(1:1)]�[0m: AHCI Port Handler: IRQ 11
> ...snip...
> �[34;1m[init_stage2(1:1)]�[0m: AHCI Port 1: Device found, Capacity=494927872, Bytes per logical sector=512, Bytes per physical sector=512
> �[34;1m[#0 colonel(0:0)]�[0m: AHCI Port Handler: IRQ received
> �[34;1m[#0 colonel(0:0)]�[0m: AHCI Port Handler: Handling IRQ for port 0
> �[34;1m[#0 colonel(0:0)]�[0m: AHCI Port 1: Interrupt handled, PxIS 1
> ```
> 
> VMware:
> 
> ```
> �[34;1m[#0 init_stage2(1:1)]�[0m: AHCI Port Handler: IRQ 7
> ...snip...
> �[34;1m[init_stage2(1:1)]�[0m: AHCI Port 1: Device found, Capacity=494927872, Bytes per logical sector=512, Bytes per physical sector=512
> �[34;1m[#0 init_stage2(1:1)]�[0m: Spurious interrupt, vector 7
> ```
> 
> From the logs, before VB and VMW trigger an interrupt at the exact same time for AHCI. However, we throw away the interrupt as "spurious" on VMW.

We could easily set a different IRQ line (when using the PIC, as we do by default).

**Comment by supercomputer7** [comment]
@Lubrsi I'm not able to boot with QEMU when chaning the IRQ line, I think I was mistaken to think that you could change the Interrupt line value and force the device to use the changed value for asserting the right IRQ line.
I do understand why `SpuriousInterruptHandler::handle_interrupt` can't handle it now, because we don't check if IRQ7 is actually a real IRQ (should be fixed to check this).

**Comment by supercomputer7** [comment]
Closed after #5916 was merged and fixed the issue :)
